### PR TITLE
Fix missing redeemers in certificate delegation and deregistration

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Fix missing redeemers in certificate delegation and deregistration
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
   - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context
Requires:
- https://github.com/input-output-hk/cardano-api/pull/268
- https://github.com/input-output-hk/cardano-cli/pull/310
- https://github.com/input-output-hk/cardano-cli/pull/320

Fixes:
- https://github.com/input-output-hk/cardano-cli/issues/299
- https://github.com/input-output-hk/cardano-cli/issues/297

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

### CODE REVIEW

**Comment by carbolymer** [comment]
~~This PR is not building, because it requires patterns exports available in cardano-api-8.23, which is introduced in: https://github.com/input-output-hk/cardano-cli/pull/320~~

File: cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs [hunk 1]
```diff
@@ -1,5 +1,3 @@
-{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs:1]: :+1:

File: cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs [hunk 2]
```diff
@@ -270,30 +269,50 @@ validateTxCertificates era certsAndScriptWitnesses = cardanoEraConstraints era $
...
-       Nothing -> Just (sCred, KeyWitness KeyWitnessForStakeAddr)
+    -- We get the stake credential witness for a certificate that requires it.
+    -- NB: Only stake address deregistration and delegation requires
+    -- witnessing (witness can be script or key)
+    deriveStakeCredentialWitness
+      :: Certificate era
+      -> Maybe StakeCredential
+    deriveStakeCredentialWitness = fmap fromShelleyStakeCredential . \case
+      ShelleyRelatedCertificate stbEra shelleyCert -> shelleyToBabbageEraConstraints stbEra $
+        case shelleyCert of
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs:257]: Don't we have this function in `cardano-api`?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

File: cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs (hunk 2 above)
Comment (carbolymer) [reply cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs:257]: These are similar ones, but are not the same: https://github.com/input-output-hk/cardano-api/blob/151eff181cb1782f9db46dbea36d22e8d40f0545/cardano-api/internal/Cardano/Api/Certificate.hs#L670

Should we extract stake creds only for deregistration and delegation like the comment says here, or should we use stake credential witness for the cases when it's available?

Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Types/Errors/TxValidationError.hs:257]: We can leave it as it for now. We should follow the specification and document it here

**Review by Jimbo4350 — APPROVED** [review]

**Pull request merged**